### PR TITLE
Introduce `VirtualHostOption` and `RouteOption` CRDs

### PR DESCRIPTION
**Pull request by saiskee**
# Description

Multiple people in the Gloo Edge community have asked for a way to externalize `options` configuration on the VirtualHost and Route level. Doing this has multiple benefits:
1. Options config (such as transformations, [headerManipulation](https://docs.solo.io/gloo-edge/master/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/headers/headers.proto.sk/#headermanipulation)) can be re-used between multiple VirtualHosts.
2. With the introduction of more complicated/verbose options config, such as Inja or XSLT transformations, VirtualService CRDs and RouteTable CRDs can become quite long and noisy. Externalizing the config to a separate CR and delegating to that CR keeps the VirtualService/RouteTable yamls lean.  

# Note

This PR only introduces the API for the Custom Resources and does not implement any logic for delegation.

### CODE REVIEW

File: projects/gateway/api/v1/virtual_service.proto [hunk 1]
```diff
@@ -168,6 +170,16 @@ message VirtualHost {
...
     gloo.solo.io.VirtualHostOptions options = 4;
+
+  // Delegate the VirtualHost options to an external VirtualHostOption Resource.
+  // Any options configured in the VirtualHost's `options` field will override all
+  // delegated options. If multiple VirtualHostOption CRs are delegated to, configuration will
+  // be taken from prior VirtualHostOption CRs over later ones.
+  // For example if `headerManipulation` is specified on the VirtualHost options, a delegated
+  // `VirtualHostOption` vhost-opt-1, and a second delegated `VirtualHostOption` vhost-opt-2, the `headerManipulation`
+  // config from only the VirtualHost-level `options` will be applied. If the config is removed from the VirtualHost-level `options` field,
+  // then the config from the first delegated `VirtualHostOption`, vhost-opt-1, is applied.
```
Comment (saiskee) [inline projects/gateway/api/v1/virtual_service.proto:181]: Wasn't sure if I should include an inline yaml example here as it won't render well in our proto docs, so I opted for an explanation.

File: test/kube2e/helm/helm_test.go [hunk 2]
```diff
@@ -44,6 +50,20 @@ var _ = Describe("Kube2e: helm", func() {
 			return string(outputBytes)
 		}, "5s", "1s").Should(ContainSubstring(rlcCrdName))
 
+		By("apply new `VirtualHostOption` CRD")
```
Comment (sam-heilbron) [inline test/kube2e/helm/helm_test.go:53]: Are these used in the test?

File: projects/gloo/api/v1/options.proto [hunk 3]
```diff
@@ -157,7 +157,7 @@ message TcpListenerOptions {
 }
 
 // Optional, feature-specific configuration that lives on virtual hosts.
-// Each VirtualHostPlugin object contains configuration for a specific feature.
+// Each VirtualHostOptions object contains configuration for a specific feature.
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/options.proto:160]: Is this change intentional? I thought the comment is in fact talking about plugins

File: projects/gateway/api/v1/virtual_service.proto [hunk 4]
```diff
@@ -218,6 +230,17 @@ message Route {
...
+
+    // Delegate the Route options to an external RouteOption Resource.
+    // Any options configured in the Route's `options` field will override all
+    // delegated options. If multiple RouteOption CRs are delegated to, configuration will
+    // be taken from prior RouteOption CRs over later ones.
+    // For example if `headerManipulation` is specified on the route options, a delegated
+    // `RouteOption` route-opt-1, and a second delegated `RouteOption` route-opt-2, the `headerManipulation`
+    // config from only the Route-level `options` will be applied. If the config is removed from the Route-level `options` field,
+    // then the config from the first delegated `RouteOption`, route-opt-1, is applied.
+    gateway.solo.io.RouteOptionRefs delegate_option = 10;
```
Comment (sam-heilbron) [inline projects/gateway/api/v1/virtual_service.proto:242]: Is there a reason we don't just use `repeated core.solo.io.ResourceRef` directly here?

File: projects/gateway/api/v1/virtual_service.proto [hunk 5]
```diff
@@ -218,6 +230,17 @@ message Route {
 
     // The name provides a convenience for users to be able to refer to a route by name.
     string name = 7;
+
+    // Delegate the Route options to an external RouteOption Resource.
+    // Any options configured in the Route's `options` field will override all
+    // delegated options. If multiple RouteOption CRs are delegated to, configuration will
+    // be taken from prior RouteOption CRs over later ones.
+    // For example if `headerManipulation` is specified on the route options, a delegated
+    // `RouteOption` route-opt-1, and a second delegated `RouteOption` route-opt-2, the `headerManipulation`
```
Comment (sam-heilbron) [inline projects/gateway/api/v1/virtual_service.proto:239]: I think an inline example could be pretty useful in addition. It's hard to tell exactly how multiple delegated refs play with an inline definition.

File: projects/gateway/api/v1/virtual_service.proto (hunk 1 above)
Comment (sam-heilbron) [reply projects/gateway/api/v1/virtual_service.proto:181]: Commented elsewhere before I saw this, but I think both inline example and explanation would be useful. We include the inline exampes with other protos

**Review by sam-heilbron — COMMENTED** [review]
This is looking good! 
Why are these being added as Gateway resources? They wrap Gloo types, so is there a particular reason they're in the gateway project?

File: install/helm/gloo/crds/gloo.solo.io_v1_RouteOption.yaml [hunk 6]
```diff
@@ -0,0 +1,3238 @@
+# Code generated by solo-kit. DO NOT EDIT.
+
+apiVersion: apiextensions.k8s.io/v1beta1
```
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply install/helm/gloo/crds/gloo.solo.io_v1_RouteOption.yaml]: We'll need to do some work to support v1 CRDs outside of this work. At the moment we read the file into a type, and if the apiVersion doesn't match, we fail: https://github.com/solo-io/solo-kit/blob/master/pkg/code-generator/schemagen/crd.go#L97. There are some other slight differences between v1beta1 and v1 that we can resolve in a wholistic approach to updating all our resources at once

File: test/kube2e/helm/helm_test.go (hunk 2 above)
Comment (saiskee) [reply test/kube2e/helm/helm_test.go:53]: no, but the CRDs need to be installed for `gateway` to create the kube clients.

File: projects/gloo/api/v1/options.proto (hunk 3 above)
Comment (saiskee) [reply projects/gloo/api/v1/options.proto:160]: I believe `VirtualHostPlugin` was renamed to `VirtualHostOptions`. There is no proto struct anymore called VirtualHostPlugin

File: projects/gateway/api/v1/virtual_service.proto (hunk 4 above)
Comment (saiskee) [reply projects/gateway/api/v1/virtual_service.proto:242]: I copied the pattern from [here](https://github.com/solo-io/gloo/blob/c67e8d72a5fb37a000d05b5d63e9d77754ea7fd1/projects/gloo/api/v1/options.proto#L197), but I now realize that this was done differently because `rate_limit_configs` is in a `oneof`.

**Comment by saiskee** [comment]
> Why are these being added as Gateway resources? They wrap Gloo types, so is there a particular reason they're in the gateway project?

The gateway project is responsible for creating the Proxy CRD, which includes all route/vhost options. This is part of the gateway translation loop and not gloo translation loop.

File: projects/gloo/api/v1/options.proto (hunk 3 above)
Comment (sam-heilbron) [reply projects/gloo/api/v1/options.proto:160]: I thought it was referring to: https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/plugins/plugin_interface.go#L163

File: install/helm/gloo/crds/gateway.solo.io_v1_RouteOption.yaml [hunk 7]
```diff
@@ -0,0 +1,3221 @@
+# Code generated by solo-kit. DO NOT EDIT.
+
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
```
Comment (sam-heilbron) [inline install/helm/gloo/crds/gateway.solo.io_v1_RouteOption.yaml:6]: We need the helm hook (https://github.com/solo-io/gloo/blob/master/install/helm/gloo/crds/gateway.solo.io_v1_Gateway.yaml#L7)

File: projects/gateway/api/v1/external_options.proto [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+*             headers:
+*               x-header-added-in-opt2:
+*                 value: this header was added in the VirtualHostOption object - #2
+* ```
+*
+* Notice how the order of VirtualHostOption delegations matters, and that the VirtualHost-level config overrides all delegated configs.
+*/
+message VirtualHostOption {
+
+  option (core.solo.io.resource).short_name = "vsopts";
```
Comment (kdorosh) [inline projects/gateway/api/v1/external_options.proto:110]: vhopts?

File: projects/gateway/api/v1/external_options.proto [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+* Notice how the order of RouteOption delegations matters, and that the Route-level option config overrides all delegated option configs.
+*/
+message RouteOption {
+
+  option (core.solo.io.resource).short_name = "rtopts";
+  option (core.solo.io.resource).plural_name = "route_options";
+
+
+  // Status indicates the validation status of this resource.
+  // Status is read-only by clients, and set by gloo during validation
```
Comment (kdorosh) [inline projects/gateway/api/v1/external_options.proto:225]: this will probably be set by gateway during validation

File: projects/gateway/api/v1/external_options.proto [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+/*
+* The **VirtualHostOption** holds `options` configuration for a VirtualHost.
+* VirtualHosts can inherit `options` config from `VirtualHostOption` objects by delegating to them.
+*
+* When a VirtualHost delegates to an external VirtualHostOptions object, any options config
+* defined on the VirtualHost will override the external options config.
+* Similarly, `VirtualHostOption`s which are delegated to first will get priority over following `VirtualHostOption`s.
+*
+* An example configuration with a VirtualService with it's own options config and delegating to two VirtualHostOption objects:
```
Comment (kdorosh) [inline projects/gateway/api/v1/external_options.proto:24]: super nit,
```suggestion
* An example configuration with a VirtualService with its own options config and delegating to two VirtualHostOption objects:Write<br class="Apple-interchange-newline"
```

File: projects/gateway/api/v1/virtual_service.proto [hunk 11]
```diff
@@ -218,6 +228,18 @@ message Route {
...
+     Any options configured in the Route's `options` field will override all
+     delegated options. If multiple RouteOption CRs are delegated to, configuration will
+     be taken from prior RouteOption CRs over later ones.
+     For example if `headerManipulation` is specified on the route options, a delegated
+     `RouteOption` route-opt-1, and a second delegated `RouteOption` route-opt-2, the `headerManipulation`
+     config from only the Route-level `options` will be applied. If the config is removed from the Route-level `options` field,
+     then the config from the first delegated `RouteOption`, route-opt-1, is applied.
+     */
+    repeated core.solo.io.ResourceRef delegate_options = 10;
+
```
Comment (kdorosh) [inline projects/gateway/api/v1/virtual_service.proto:242]: thoughts on putting these delegate options into their own message? we might want to make this a oneof in the future.

i'm imaging a v2 iteration on this feature with more complex selection logic, (e.g. k8s labels) akin to what we have today for route table delegation

File: projects/gateway/api/v1/external_options.proto [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+*       maxAge: 2s
+*     transformations:
+*       requestTransformation:
+*         transformationTemplate:
+*           headers:
+*             x-header-added-in-opt2:
+*               value: this header was added in the VirtualHostOption object - #2
+* ```
+*
+* The final virtual host options (visible in the Proxy CRD) would be:
```
Comment (kdorosh) [inline projects/gateway/api/v1/external_options.proto:83]: super nit,
```suggestion
* The final virtual host options (visible in the Proxy CR) would be:
```

**Review by kdorosh — COMMENTED** [review]
looks good, just nits and the follow up on future-proofing the API in case we want to add other CR delegation selectors

File: projects/gateway/api/v1/virtual_service.proto (hunk 11 above)
Comment (saiskee) [reply projects/gateway/api/v1/virtual_service.proto:242]: This sounds good to me! I wasn't sure if we would want to introduce more complex selection logic in the future, but might as well include the `oneof` to prevent any breaks in the future.

**Review by kdorosh — COMMENTED** [review]
lgtm, waiting for https://github.com/solo-io/gloo/pull/4868 to merge in

**Review by kdorosh — DISMISSED** [review]

**Review by sam-heilbron — APPROVED** [review]
